### PR TITLE
Add time estimation

### PR DESCRIPTION
**Pull request by mr-field**
### Description of work
Adds a new column to the Script Generator table to show the estimated time in seconds necessary to complete each action. The column is populated only if the action is valid, otherwise it will be empty. The PR also adds a label at the bottom of the table to display the total estimated run time.

The implementation is very similar to the `parameters_valid` function, but in this case the script definition should return an `int` value representing the estimate. If the value is actually a float, it will be rounded to the nearest integer before being passed to Java. If it's not a numeric type, it will be ignored.

Note that Py4J coverts a Python `int` to a Java `Integer` or `Long` depending on its size. To avoid casting exceptions when handling large estimates, the wrapper classes use the `Number` abstract class to hold the value returned by Py4J; when it needs to be used in a calculation, it's converted to `long`.

**Note**: This needs Squish tests, but I won't be in the group for long enough to write them.

### Ticket
https://github.com/ISISComputingGroup/IBEX/issues/4172

### Acceptance criteria
- The new Python unit tests pass
- A new Script Generator column is displayed for time esimates
- The column updates correctly when the action parameters are edited, as specified in the ScriptDefinitions's `estimate_time` function
- No estimate is displayed if parameters are invalid or `estimate_time` is not implemented
- A new label is displayed under the action table showing the current total estimated time
- The label is updated correctly when actions are edited or removed

### Unit tests
Some tests for the Python generator have been added to `uk.ac.stfc.isis.ibex.scriptgenerator\python_support`. They can be run in Genie Python 3 with `python -m unittest test_generator.py`

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by DominicOram** [comment]
Great work @mr-field, works really well! Just some tidying up that I would like to do. Apologies it's taken me so long to get round to looking at it. General first impressions that I'm putting here to remind me before I do a more in depth review:
* `None` showing blank is not very user friendly
* The script generator view will crash if time estimate errors e.g. enter `1e6` in the `example1` script
* I think changing the formatting when rows are in the order of minutes/hours would be good

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorViewModel.java [hunk 1]
```diff
@@ -660,6 +694,31 @@ public String getToolTipText(Object element) {
...
+                new DataboundCellLabelProvider<ScriptGeneratorAction>(viewTable.observeProperty("time estimate")) {
+            @Override
+            protected String stringFromRow(ScriptGeneratorAction row) {
+                if (!scriptGeneratorModel.languageSupported) {
+                    return "\u003F"; // A question mark to say we cannot be certain
+                }
+                
+                Optional<Number> estimatedTime = row.getEstimatedTime();
+                if (estimatedTime.isEmpty()) {
+                    return "";
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorViewModel.java:709]: Might be better to say `Unknown`?

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/pythoninterface/PythonInterface.java [hunk 2]
```diff
@@ -405,6 +411,39 @@ public void refreshAreParamsValid(List<ScriptGeneratorAction> scriptGenContent,
...
+            }, THREAD).thenAccept(timeEstimate -> 
+                firePropertyChange(TIME_ESTIMATE_PROPERTY, null, timeEstimate)
+            );
+        } else {
+            handlePythonReadinessChange(false);
+            throw new PythonNotReadyException("When getting time estimation");
+        }
+        if (!pythonReady) {
+            throw new PythonNotReadyException("When getting time estimation");
+        }
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/pythoninterface/PythonInterface.java:445]: There are also 3 or 4 almost identical functions in this class, can we pull out the similarities?

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java [hunk 3]
```diff
@@ -576,6 +589,33 @@ public void refreshParameterValidityChecking() throws NoScriptDefinitionSelected
...
+                .orElseThrow(() -> new NoScriptDefinitionSelectedException(
+                        "Tried to refresh time estimation with no script definition selected"));
+        try {
+            generator.refreshTimeEstimation(scriptGeneratorTable, scriptDefinition);
+            languageSupported = true;
+            threadError = false;
+        } catch (UnsupportedLanguageException e) {
+            firePropertyChange(LANGUAGE_SUPPORT_PROPERTY, languageSupported, false);
+            LOG.error(e);
+            languageSupported = false;
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java:612]:
```suggestion
            firePropertyChange(LANGUAGE_SUPPORT_PROPERTY, languageSupported, languageSupported=false);
            LOG.error(e);
```
Similar for thread error below.

Comment (DominicOram) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java:612]: This same pattern of error handling is used in lots of places in this class, can we generalise it?

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java [hunk 4]
```diff
@@ -128,6 +128,8 @@
 	 * all script generator contents are valid or not (bool).
 	 */
 	private static final String PARAM_VALIDITY_PROPERTY = "parameter validity";
+	
+	private static final String TIME_ESTIMATE_PROPERTY = "time estimate";
```
[the remark this answers is not in the document]
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java:132]: Would it be useful to have all constants in one file? Seems like there are other constants defined the same number of times for a script generator

[the remark this answers is not in the document]
Comment (DominicOram) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java:132]: Probably, but I'm not sure where makes the most sense

**Pull request merged**